stellargraph/datasets/dataset_loader.py

```diff
@@ -0,0 +1,195 @@
+    name = ""
+    directory_name = ""
+    url = ""
+    url_archive_format: Optional[str] = None
+    expected_files: List[str] = []
+    description = ""
+    source = ""
+    data_subdirectory_name: Optional[str] = None
+
+    @classmethod
+    def __init_subclass__(
+        cls,
+        name: str,
+        directory_name: str,
+        url: str,
+        url_archive_format: Optional[str],
+        expected_files: List[str],
+        description: str,
+        source: str,
+        data_subdirectory_name: Optional[str] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Used to set class variables during the class definition of derived classes and generate customised docs.
+        NOTE: this is not compatible with python's ABC abstract base class, so this class derives from object."""
+        cls.name = name
+        cls.directory_name = directory_name
+        cls.url = url
+        cls.url_archive_format = url_archive_format
+        cls.expected_files = expected_files
+        cls.description = description
+        cls.source = source
+        cls.data_subdirectory_name = data_subdirectory_name
+
+        if url_archive_format is None and len(expected_files) != 1:
+            raise ValueError(
+                "url_archive_format is None, which requires a single expected_file, found: {expected_files!r}"
+            )
+
+        # auto generate documentation
+        if cls.__doc__ is not None:
+            raise ValueError(
+                "DatasetLoader docs are automatically generated and should be empty"
+            )
+        cls.__doc__ = f"{cls.description}\n\nFurther details at: {cls.source}"
+
+        super().__init_subclass__(**kwargs)  # type: ignore
+
+    def __init__(self) -> None:
+        # basic check since this is effectively an abstract base class, and derived classes should have set name
+        if not self.name:
+            raise ValueError(
+                f"{self.__class__.__name__} can't be instantiated directly, please use a derived class"
+            )
+
+    @property
+    def base_directory(self) -> str:
+        """str: The full path of the directory containing this dataset."""
+        return os.path.join(self._all_datasets_directory(), self.directory_name)
+
+    @property
+    def data_directory(self) -> str:
+        """str: The full path of the directory containing the data content files for this dataset."""
+        if self.data_subdirectory_name is None:
+            return self.base_directory
+        else:
+            return os.path.join(self.base_directory, self.data_subdirectory_name)
+
+    def _create_base_directory(self) -> None:
+        data_dir = self.base_directory
+        if not os.path.exists(data_dir):
+            os.makedirs(data_dir)
+
+    @staticmethod
+    def _all_datasets_directory() -> str:
+        """Return the path of the base directory which contains subdirectories for each dataset."""
+        return os.path.expanduser(
```

Comment:
Very minor thing, but I don't think the envvar should be expanded, strictly speaking, because the shell will do that:

```bash
$ X=~/foo
$ echo $X
/Users/.../foo
```

By expanding twice here we risk overruling the users intention, if they wanted to write to a directory that literally started with `~/`:

```bash
$ export X=\~
$ echo $X
~
$ python -c 'import os; print(os.path.expanduser(os.environ["X"]))'
/Users/wil9dw
```

That's very much an edge case, but:

```python
os.getenv("STELLARGRAPH_DATASETS_PATH", os.path.expanduser("~/stellargraph-datasets"))
```

(FWIW, I think `/` is fine as a literal directory separator, even on windows: https://docs.microsoft.com/en-us/dotnet/standard/io/file-path-formats#canonicalizing-separators.)